packages/firestore/src/local/local_store.ts

```diff
@@ -702,41 +702,40 @@ export class LocalStore {
    * Notify local store of the changed views to locally pin documents.
    */
   notifyLocalViewChanges(viewChanges: LocalViewChanges[]): Promise<void> {
+    for (const viewChange of viewChanges) {
+      const targetId = viewChange.targetId;
+
+      this.localViewReferences.addReferences(viewChange.addedKeys, targetId);
+      this.localViewReferences.removeReferences(
+        viewChange.removedKeys,
+        targetId
+      );
+
+      if (!viewChange.fromCache) {
+        const queryData = this.queryDataByTarget.get(targetId);
+        assert(
+          queryData !== null,
+          `Can't set limbo-free snapshot version for unknown target: ${targetId}`
+        );
+
+        // Advance the last limbo free snapshot version
+        const lastLimboFreeSnapshotVersion = queryData!.snapshotVersion;
+        const updatedQueryData = queryData!.withLastLimboFreeSnapshotVersion(
+          lastLimboFreeSnapshotVersion
+        );
+        this.queryDataByTarget = this.queryDataByTarget.insert(
+          targetId,
+          updatedQueryData
+        );
+      }
+    }
     return this.persistence.runTransaction(
       'notifyLocalViewChanges',
       'readwrite',
```

Comment:
Here and below: Can you change this to `readwrite-idempotent`? Note that if you merge in the latest changes from mrschmidt/idempotent, the test runner will automatically run the transaction twice (so you can verify that this works).

Reply:
Done.